server/confdata-binlog-replay.cpp

```diff
@@ -44,6 +65,40 @@ class ConfdataBinlogReplayer : vk::binlog::replayer {
 
   using vk::binlog::replayer::replay;
 
+  // To prevent OOM errors on confdata overflows we start to throttle events, when too little shared memory left:
+  // Soft OOM (85%) -- replay only existing keys related events until server restart
+  // Hard OOM (95%) -- ignore all events, don't read binlog at all until server restart
+  enum class MemoryStatus {
+    NORMAL,
+    SOFT_OOM,
+    HARD_OOM
+  };
+
+  MemoryStatus current_memory_status() noexcept {
+    update_memory_status();
+    if (hard_oom_reached_) return MemoryStatus::HARD_OOM;
+    if (soft_oom_reached_) return MemoryStatus::SOFT_OOM;
+    return MemoryStatus::NORMAL;
+  }
+
+  bool check_has_enough_memory(size_t need_bytes, std::string_view msg) noexcept {
+    if (memory_resource_->is_enough_memory_for(need_bytes)) {
+      return true;
+    }
+    log_server_critical("Not enough confdata shared memory. Processing key with first part = '%s'. %zu bytes needed by estimation for %s",
+                        processing_key_.get_first_key().c_str(), need_bytes, msg.data());
```

Comment:
In general, `.data()` is incorrect for `string_view`, since it may not end with `\0`. Probably, the easiest way is just to use `const char *`, as below

Reply:
replaced to `const char *`